### PR TITLE
added reconcile iterator

### PR DESCRIPTION
**Pull request by john-hill**

### CODE REVIEW

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/replication/ReconcileStreamUtilImpl.java [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+import java.util.Iterator;
+
+import org.sagebionetworks.repo.model.IdAndChecksum;
+import org.sagebionetworks.repo.model.ObjectType;
+import org.sagebionetworks.repo.model.message.ChangeMessage;
+import org.springframework.stereotype.Service;
+
+@Service
+public class ReconcileStreamUtilImpl implements ReconcileStreamUtil {
```
Comment (marcomarasca) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/replication/ReconcileStreamUtilImpl.java]: We might want to move this to the manager as this layer of abstraction is kind of redundant

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/replication/ReconcileIterator.java [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+		if (replication.hasNext()) {
+			return Optional.of(replication.next());
+		} else {
+			return Optional.empty();
+		}
+	}
+
+	private Iterator<ChangeMessage> findNextBatch() {
+		List<ChangeMessage> results = new ArrayList<>();
+		Optional<IdAndChecksum> t = truthNext();
```
Comment (marcomarasca) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/replication/ReconcileIterator.java]: t -> truthItem

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/replication/ReconcileIterator.java [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+		Optional<IdAndChecksum> r = replicationNext();
```
Comment (marcomarasca) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/replication/ReconcileIterator.java]: r -> replicationItem

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/replication/ReconcileIterator.java [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+							}
+						}
+					}
+				}
+			} else {
+				// truth is exhausted
+				if(r.isPresent()) {
+					results.add(newChange(ChangeType.DELETE, r.get().getId()));
+					return results.iterator();
+				}else {
```
Comment (marcomarasca) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/replication/ReconcileIterator.java:118]: remove return and } else

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/replication/ReconcileIterator.java [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+		Optional<IdAndChecksum> r = replicationNext();
+		while (true) {
+			if (t.isPresent()) {
+				if (!r.isPresent()) {
+					// object missing from replication
+					results.add(newChange(ChangeType.CREATE, t.get().getId()));
+					return results.iterator();
+				} else {
+					// both t and r exist
+					IdAndChecksum fromT = t.get();
```
Comment (marcomarasca) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/replication/ReconcileIterator.java]: Let us get rid of the Optional as it complicates the code for no reason

File: services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/replication/ReconcileIteratorTest.java [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+				new ChangeMessage().setObjectType(ObjectType.ENTITY).setChangeType(ChangeType.UPDATE).setObjectId("1"),
+				new ChangeMessage().setObjectType(ObjectType.ENTITY).setChangeType(ChangeType.UPDATE).setObjectId("3")
+		);
+		assertEquals(expected, results);
+	}
+	
+	@Test
+	public void testIteratorWithSameRunWithNoMatches() {
+		List<IdAndChecksum> truth = Arrays.asList(
+				new IdAndChecksum().withId(1L).withhecksum(0L),
```
Comment (marcomarasca) [inline services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/replication/ReconcileIteratorTest.java]: Let us have a case same as this one where truth and rep are flipped.

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/replication/ReconcileIterator.java [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+ * Note: Both streams are expected to be ordered by ID ascending.
+ *
+ */
+public class ReconcileIterator implements Iterator<ChangeMessage> {
+
+	private final Iterator<IdAndChecksum> truth;
+	private final Iterator<IdAndChecksum> replication;
+	private final ObjectType objectType;
+
+	Iterator<ChangeMessage> currentBatch;
```
Comment (marcomarasca) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/replication/ReconcileIterator.java]: private

File: lib/models/src/main/java/org/sagebionetworks/repo/model/IdAndChecksum.java [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+	/**
+	 * @return the checksum
+	 */
+	public Long getChecksum() {
+		return checksum;
+	}
+	/**
+	 * @param checksum the checksum to set
+	 */
+	public IdAndChecksum withhecksum(Long checksum) {
```
Comment (marcomarasca) [inline lib/models/src/main/java/org/sagebionetworks/repo/model/IdAndChecksum.java]: withhecksum -> withChecksum

**Review by marcomarasca — APPROVED** [review]

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/replication/ReconcileIterator.java (hunk 4 above)
Comment (john-hill) [reply services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/replication/ReconcileIterator.java:118]: I purposely kept each if and else for book keeping.

**Review by marcomarasca — APPROVED** [review]

**Pull request merged**